tensorstore/internal/http/http_request.h

```diff
@@ -62,17 +70,22 @@ class HttpRequest {
 /// Implements the builder pattern for HttpRequest.
 class HttpRequestBuilder {
  public:
-  /// Creates a request builder, using the specified method and url.
+  using UriEncodeFunctor = absl::FunctionRef<std::string(std::string_view)>;
+  /// Creates a request builder, using the specified method, url and URI encoding function.
   ///
   /// The method should be an HTTP method, like "GET", "POST", "PUT", "HEAD",
   /// etc.
-  explicit HttpRequestBuilder(std::string_view method, std::string base_url);
+  explicit HttpRequestBuilder(std::string_view method, std::string base_url)
+    : HttpRequestBuilder(method, base_url,
+        [](std::string_view d) { return internal::PercentEncodeUriComponent(d); }) {}
+
+  explicit HttpRequestBuilder(std::string_view method, std::string base_url, UriEncodeFunctor uri_encoder);
 
   /// Creates an http request with the given payload.
   ///
   /// This function invalidates the builder. The application should not use this
   /// builder once this function is called.
-  HttpRequest BuildRequest();
+  Result<HttpRequest> BuildRequest();
```

Comment:
This change to `Result<HttpRequest>` and the addition of the `status_` member can be reverted.  The two cases where an error status can occur, `HttpRequestBuilder::MaybeAddRangeHeader` and `S3RequestBuilder::AddHeader`, are both cases where the caller should instead be responsible for ensuring that the byte range/header is valid, and those checks can just be asserts rather than error returns.